From 71f97948230c6e4f9fb7444477893fa73e4fa9ed Mon Sep 17 00:00:00 2001
From: Don Zickus <dzickus@redhat.com>
Date: Tue, 16 Sep 2014 17:07:54 -0400
Subject: [usb] catch attempts to submit urbs with a vmalloc'd transfer buffer

Message-id: <1410887463-20674-403-git-send-email-dzickus@redhat.com>
Patchwork-id: 95339
O-Subject: [RHEL7 PATCH 402/591] usb: catch attempts to submit urbs with a vmalloc'd transfer buffer
Bugzilla: 1110939
RH-Acked-by: Tony Camuso <tcamuso@redhat.com>
RH-Acked-by: Prarit Bhargava <prarit@redhat.com>

https://bugzilla.redhat.com/show_bug.cgi?id=1110939

commit 29d2fef8be1165a26984a94fbcf81d68c1442fc5
Author: Dan Williams <dan.j.williams@intel.com>
Date:   Thu May 8 19:25:56 2014 +0300

    usb: catch attempts to submit urbs with a vmalloc'd transfer buffer

    Save someone else the debug cycles of figuring out why a driver's
    transfer request is failing or causing undefined system behavior.
    Buffers submitted for dma must come from GFP allocated / DMA-able
    memory.

    Return -EAGAIN matching the return value for dma_mapping_error() cases.

    Acked-by: Alan Stern <stern@rowland.harvard.edu>
    Cc: Sarah Sharp <sarah.a.sharp@linux.intel.com>
    Cc: Mathias Nyman <mathias.nyman@linux.intel.com>
    Signed-off-by: Dan Williams <dan.j.williams@intel.com>
    Signed-off-by: Mathias Nyman <mathias.nyman@linux.intel.com>
    Signed-off-by: Greg Kroah-Hartman <gregkh@linuxfoundation.org>
---

Signed-off-by: Jarod Wilson <jarod@redhat.com>

diff --git a/drivers/usb/core/hcd.c b/drivers/usb/core/hcd.c
index 3af0ab8..d4640f6 100644
--- a/drivers/usb/core/hcd.c
+++ b/drivers/usb/core/hcd.c
@@ -1502,6 +1502,9 @@ int usb_hcd_map_urb_for_dma(struct usb_hcd *hcd, struct urb *urb,
      ret = -EAGAIN;
     else
      urb->transfer_flags |= URB_DMA_MAP_PAGE;
+   } else if (is_vmalloc_addr(urb->transfer_buffer)) {
+    WARN_ONCE(1, "transfer buffer not dma capable\n");
+    ret = -EAGAIN;
    } else {
     urb->transfer_dma = dma_map_single(
       hcd->self.controller,
-- 
1.7.1